### PR TITLE
Last Features

### PR DESCRIPTION
**Pull request by KamilRosiak**
1. When the new ConstraintView is not opened you get a Dialog when opening a new FeatureModel to choose whether you want the ConstraintView to be opened or not. Also you have the opportunity to change the decision under Window -> Preferences -> FeatureIDE
2. Clicking on a Constraint that e.g. is redundant the Constraints causing this get marked with a circle. 
3. In the toolbar at the top of the View you can create a new constraint, refresh the view and Hide/Show collapsed constraints with shortcuts.
4. Added "Focus on Explanation" and "Focus on contained Features" to the Context Menu
5. Added ShortCuts to the View: 
   - Del: Deletes a Constraint
   - Ctrl+F: you enter the searchbox
   - Ctrl+Z/Ctrl+Shift+Z: Undo/Redo 
   - Esc: Unselect the selected Constraints
6. Refactored and fixed bugs a lot

### CODE REVIEW

**Review by ThisIsATunaSandwich — COMMENTED** [review]
Works and Code looks good.
@PaulWestphal

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 1]
```diff
@@ -464,6 +465,7 @@
 	public static final String DELETE_INCLUDING_SUBFEATURES = "Delete Including Subfeatures (CTRL + D)";
 	public static final String COLLAPSE_FEATURE = "Collapse (CTRL + C)";
 	public static final String EDIT_CONSTRAINT = "Edit Constraint";
+	public static final String ADD_GROUP_TAG_TO_DESCRIPTION = "Add group tag to Description";
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java]: Inconsistent capitalization

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureModelEditor.java [hunk 2]
```diff
@@ -108,6 +108,7 @@
  *
  * @author Thomas Thuem
  * @author Christian Becker
+ *
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/FeatureModelEditor.java]: ?

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ShowCollapsedConstraintsAction.java [hunk 3]
```diff
@@ -63,5 +63,4 @@ public void run() {
 
 		}
 	}
-
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/editors/featuremodel/actions/ShowCollapsedConstraintsAction.java:66]: ?

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/FocusOnExplanationInViewAction.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+		// apply explanation collapse
+		if (focusOnExplanationOperation != null) {
+			try {
+				PlatformUI.getWorkbench().getOperationSupport().getOperationHistory().execute(focusOnExplanationOperation, null, null);
+			} catch (final ExecutionException e) {
+				FMUIPlugin.getDefault().logError(e);
+			}
+		}
+	}
+
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/FocusOnExplanationInViewAction.java:100]: Simplify method

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+				}
+				shell.dispose();
+			}
+		});
+		cancelButton.setLayoutData(rowData);
+		yesButton.setLayoutData(rowData);
+		noButton.setLayoutData(rowData);
+		shell.setDefaultButton(cancelButton);
+	}
+
```
Comment (tthuem) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java]: Simplify/split method

**Comment by tthuem** [comment]
@PaulWestphal Please test whether other views also work as before. For example, the collaboration diagram did not show any content in prior versions.

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/Preferences.java [hunk 6]
```diff
@@ -54,6 +53,21 @@ public static void setDefaultCompletion(int defaultCompletion) {
...
+	/**
+	 * this method adds a value by his key to the preferences
+	 */
+	public static void store(String key, String value) {
+		preferences.put(key, value);
+		flush();
+	}
+
+	/**
+	 * get the value of an given key
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/Preferences.java:65]: a* given key

File: plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java [hunk 7]
```diff
@@ -555,6 +557,10 @@
 	public static final String UNABLE_TO_START_GUIDSL = "Unable to start GUIDSL";
 	public static final String DIMACS = "Dimacs";
 	public static final String CONFIGURATION_COLORING = "Configuration Coloring";
+	public static final String CONFIGURATION_DIALOGS = "Configuration Dialogs";
+	public static final String CONFIGURATION_DIALOGS_CONSTRAINT_TEXT = "Turn off the ConstraintView Dialog.";
+	public static final String CONFIGURATION_DIALOGS_CONSTRAINT_TOOLTIP =
+		"If this option is not selected, upon opening a feature diagramm FeatureIDE will ask you if you want to open the ConstraintView. Uncheck this if you want to change the default behaviour.";
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.core/src/de/ovgu/featureide/fm/core/localization/StringTable.java:563]: Positive phrasing is generally better:

"If this option **is** selected,"

Also diagram is misspelled.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/preferences/FeatureIDEPreferencePage.java [hunk 8]
```diff
@@ -72,6 +76,7 @@ protected Control createContents(Composite parent) {
 		final Composite container = new Composite(parent, SWT.NULL);
 		container.setLayout(new FillLayout(SWT.VERTICAL));
 
+		// completion group
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/preferences/FeatureIDEPreferencePage.java:79]: Pointless

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/DeleteConstraintAction.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+import de.ovgu.featureide.fm.ui.editors.featuremodel.operations.DeleteConstraintOperation;
+import de.ovgu.featureide.fm.ui.editors.featuremodel.operations.ElementDeleteOperation;
+
+/**
+ * This class represents the Action to delete one or multiply Constraints selected in the ConstraintView.
+ *
+ * @author "Rosiak Kamil"
+ * @author "Rahel Arens"
+ */
+public class DeleteConstraintAction extends Action {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/DeleteConstraintAction.java:44]: Add "InView" for consinstent naming.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/FocusOnContainedFeaturesAction.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import de.ovgu.featureide.fm.ui.FMUIPlugin;
+import de.ovgu.featureide.fm.ui.editors.IGraphicalFeatureModel;
+import de.ovgu.featureide.fm.ui.editors.featuremodel.operations.ExpandConstraintOperation;
+
+/**
+ * This class represents the Action to focus on contained Features of a constraint selected in the ConstraintView.
+ *
+ * @author Rahel Arens
+ */
+public class FocusOnContainedFeaturesAction extends Action {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/FocusOnContainedFeaturesAction.java:41]: Add InView for consistent naming.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/MinimalCreateConstraintAction.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import de.ovgu.featureide.fm.core.base.IFeatureModel;
+import de.ovgu.featureide.fm.core.localization.StringTable;
+import de.ovgu.featureide.fm.ui.editors.ConstraintDialog;
+
+/**
+ * A modified CreateConstraintAction for the ConstraintView
+ *
+ * @author Domenik Eichhorn
+ */
+public class MinimalCreateConstraintAction extends Action {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/MinimalCreateConstraintAction.java:37]: Add InView for consistent naming.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/ShowCollapsedConstraintAction.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import de.ovgu.featureide.fm.ui.FMUIPlugin;
+import de.ovgu.featureide.fm.ui.editors.IGraphicalFeatureModel;
+import de.ovgu.featureide.fm.ui.editors.featuremodel.operations.ShowCollapsedConstraintsOperation;
+
+/**
+ * A modified ShowCollapsedConstraintsAction for the ConstraintView
+ *
+ * @author Domenik Eichhorn
+ */
+public class ShowCollapsedConstraintAction extends Action {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/actions/ShowCollapsedConstraintAction.java:42]: Add InView for consistent naming. Also make constraint plural.

**Comment by PaulWestphal** [comment]
constraintview
The "would you like to open the constraint view" dialog is still displayed while eclipse is being started.

**Comment by PaulWestphal** [comment]
The other views (configuration map, collaboration diagram) seem to work again.

**Comment by tthuem** [comment]
@PaulWestphal If this is the only issue, go ahead an merge it. Most users will open Constraints View anyway and then the message will not appear.

Just to clarify: It only pops-up if the view is not opened, a feature model is opened, and remember my decision has not been checked before, right? If so, go ahead an merge it.

**Comment by PaulWestphal** [comment]
Yes, it will only pop up in that case. Another problem is that under windows, the dots indicating the reason why features are defect don't seem to show up. 

I'm also still working on the code review itself, there are some naming inconsistencies.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintColorPair.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+ * See http://featureide.cs.ovgu.de/ for further information.
+ */
+package de.ovgu.featureide.fm.ui.views.constraintview.util;
+
+import org.eclipse.swt.graphics.Color;
+
+import de.ovgu.featureide.fm.core.base.IConstraint;
+
+/**
+ * This class maps a IConstraint to an Color.
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintColorPair.java:28]: an* IConstraint to a* Color

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+		while (!shell.isDisposed()) {
+			if (!display.readAndDispatch()) {
+				display.sleep();
+			}
+		}
+		return exitCode;
+	}
+
+	/**
+	 * This method creating a label with the question.
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java:104]: creates* a label with the question.

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+		rememberDecButton.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent event) {
+				remember = rememberDecButton.getSelection();
+			}
+		});
+	}
+
+	/**
+	 * This method creating the buttons on the bottom.
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/util/ConstraintViewDialog.java:128]: creates*

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				System.out.println("Constraint Header");
+
+			}
+
+			@Override
+			public void widgetDefaultSelected(SelectionEvent e) {
+				// TODO Auto-generated method stub
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java:271]: Remove TODO

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+		});
+		descriptionColumn = new TreeColumn(viewer.getTree(), SWT.LEFT);
+		descriptionColumn.setResizable(true);
+		descriptionColumn.setMoveable(true);
+		descriptionColumn.setWidth(CONSTRAINT_DESCRIPTION_WIDTH);
+		descriptionColumn.setText(DESCRIPTION_HEADER);
+	}
+
+	/**
+	 * Text searchBox
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintView.java:283]: Add description

File: plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintViewSettingsMenu.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+		final IToolBarManager toolBarManager = actionBars.getToolBarManager();
+		toolBarManager.add(createAction);
+		toolBarManager.add(refreshAction);
+		toolBarManager.add(collapseAction);
+	}
+
+	/**
+	 * disables all actions (run method is not called when activated)
+	 */
+	public void setStatOfActions(Boolean isEnabled) {
```
Comment (PaulWestphal) [inline plugins/de.ovgu.featureide.fm.ui/src/de/ovgu/featureide/fm/ui/views/constraintview/view/ConstraintViewSettingsMenu.java:98]: setStateOfActions*

**Pull request merged**